### PR TITLE
feature(assets): Move all core static assets to views

### LINKED ISSUE #5105: Move graphics into views system
Now that we support #4476, we should move the files in _graphics into the views system. Couple reasons:

1. Cache-busting -- sites will show stale versions unless we update the filename every time we update the image. That's a minor hassle, since you also have to go in and change the version in the CSS/HTML by hand as well.

2. Overrideable -- would be awesome to be able to drop a new favicon at ```views/defaut/graphics/favicon.ico``` or something and have it "just work". Right now the process is:
 - 1. put a new favicon somewhere in your plugin
  - 2. override the ```page/elements/shortcut_icon``` view with the html needed to define the favicon, including building a URL that points to the correct location in the ```/mod/``` directory (blegh). Default group icon is even worse, since you have to register for a plugin hook or something.

### PR DESCRIPTION
**Pull request by ewinslow**
Fixes #5105

TODOs:
- [x] Add property-read for the interface
- [x] Add a section in guides/services.rst
- [x] Add clearer usage guide in the interface docs
- [x] Drop superfluous `elgg_register_simplecache_view` calls

### CODE REVIEW

File: engine/classes/Elgg/Application/CacheHandler.php [hunk 1]
```diff
@@ -199,19 +203,59 @@ protected function sendCacheHeaders($etag) {
...
+				return 'text/css';
+				
+			case 'html':
+				return 'text/html';
+				
+			case 'xml':
+				return 'text/xml';
+			
+			default:
+				break;
```
Comment (ewinslow) [inline engine/classes/Elgg/Application/CacheHandler.php]: We'll want to make this extensible probably, but for now I just hard-coded a bunch of common asset media types.

File: views/default/resources/admin/plugin_screenshot.img.php [hunk 2]
```diff
@@ -9,11 +9,11 @@
 
 $plugin = elgg_get_plugin_from_id($plugin_id);
 if (!$plugin) {
-	$file = elgg_get_root_path() . '_graphics/icons/default/medium.png';
+	$file = elgg_get_root_path() . 'views/default/icons/default/medium.png';
```
Comment (mrclay) [inline views/default/resources/admin/plugin_screenshot.img.php]: Why not asset URLs here?

Comment (ewinslow) [inline views/default/resources/admin/plugin_screenshot.img.php]: Plugin screenshot needs the file path, not the URL. I suppose I could use the url and let PHP's file handling of URLs kick in, but I didn't think of it at the time and this seemed like an easier refactoring.

File: engine/classes/Elgg/Application.php [hunk 3]
```diff
@@ -46,6 +46,7 @@ class Application {
 	 * @var string[]
 	 */
 	private static $public_services = [
+		'assets' => true,
```
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: Also needs property-read for the interface, a section in guides/services.rst, and clearer guide to usage in the interface.

Comment (ewinslow) [inline engine/classes/Elgg/Application.php]: Thanks for keeping me honest! Will get to those things asap...

**Comment by mrclay** [comment]
This some big performance implications and should get a thumbs up from a few core devs. I'm hesitantly behind the idea, though it doesn't seem critical IMO.

**Comment by ewinslow** [comment]
@mrclay I thought we'd agreed that there were other solutions for folks who are truly concerned about the perf implications here?

In any case, I'm fine with waiting for other LGTMs from @Elgg/owners that have strong opinions on this. Specifically maybe @Srokap IIRC.

**Comment by ewinslow** [comment]
Changes made. Perhaps @Elgg/contributors would like to weigh in too.

**Comment by mrclay** [comment]
It says the assets svc is to "effectively replace `elgg_get_simplecache_url()`". Can that existing function already be used for this? If given the option of not adding 2 more classes and more API, I'd rather do that.

**Comment by ewinslow** [comment]
Good question. It can't be used because it can't handle arbitrary assets at
arbitrary locations in the views system. It only handles JS and CSS in the
js/ and css/ directories.

Assets::getUrl allows you to reference images, and to do so outside the js
and css directories. (E.g. "favicon.ico")

**Comment by mrclay** [comment]
We can't just allow `""` for the first argument?

Steve
-- 
http://community.elgg.org/profile/steve_clay

**Comment by mrclay** [comment]
...or detect if the js/css arg isn't given.

**Comment by ewinslow** [comment]
Yes, we could do that.

File: engine/classes/Elgg/Cache/SimpleCache.php [hunk 4]
```diff
@@ -49,30 +49,50 @@ function registerView($view_name) {
...
+		
+		// handle `getUrl('js', 'blog/save_draft.js')`
+		if (!empty($subview)) {
+			$view = "$view/$subview";
+			$subview = '';
 		}
 	
-		elgg_register_simplecache_view("$type/$view");
-		return _elgg_get_simplecache_root() . "$type/$view";
+		// should be normalized to canonical form by now: `getUrl('js/blog/save_draft.js')`
```
Comment (mrclay) [inline engine/classes/Elgg/Cache/SimpleCache.php]: :+1:

File: engine/classes/Elgg/Di/ServiceProvider.php [hunk 5]
```diff
@@ -71,21 +71,6 @@ class ServiceProvider extends \Elgg\Di\DiContainer {
 	 */
 	public function __construct(\Elgg\Config $config) {
 
-		$this->setFactory('classLoader', function(ServiceProvider $c) {
```
Comment (mrclay) [inline engine/classes/Elgg/Di/ServiceProvider.php]: May I gently suggest leaving this reordering out of this PR?

File: engine/lib/cache.php [hunk 6]
```diff
@@ -100,22 +100,34 @@ function elgg_register_simplecache_view($view_name) {
...
- *		elgg_load_js('elgg.blog');
+ * Note that with this older approach, there is no way to access cached views
+ * outside of the js/ or css/ folders.
  *
- * @param string $type The file type: css or js
- * @param string $view The view name after css/ or js/
+ * This automatically registers the view with Elgg's simplecache.
+ * 
+ * @param string $view The full view name
+ * @param string $subview If the first arg is css/ or js/, the rest of the view name
```
Comment (mrclay) [inline engine/lib/cache.php]: `If the first arg is "css" or "js"` (we don't allow trailing slash)

File: engine/classes/Elgg/Application/CacheHandler.php (hunk 1 above)
Comment (mrclay) [inline engine/classes/Elgg/Application/CacheHandler.php]: I'm OK with this.

File: docs/guides/views/simplecache.rst [hunk 7]
```diff
@@ -38,13 +40,11 @@ You can register a view with the Simplecache with the following function at init
...
 .. code:: php
 
-   $url = elgg_get_simplecache_url($type, $view)
-
-Where:
-
-- ``$type`` is js or css
-- ``$view`` the view name after css/ or js/
\ No newline at end of file
+   $js = elgg_get_simplecache_url('js/your_view.js')
```
Comment (mrclay) [inline docs/guides/views/simplecache.rst]: semicolons :)

**Comment by mrclay** [comment]
Is simplecache going to know to JSmin views like `foo/bar.js`? If not, please make a separate issue for that.

**Comment by mrclay** [comment]
Do we already automatically register a view `foo.js` for external? I know we do for `js/foo.js`.

File: engine/classes/Elgg/Di/ServiceProvider.php (hunk 5 above)
Comment (ewinslow) [inline engine/classes/Elgg/Di/ServiceProvider.php]: yup, good call

File: docs/guides/views/simplecache.rst (hunk 7 above)
Comment (ewinslow) [inline docs/guides/views/simplecache.rst]: done

**Comment by ewinslow** [comment]
yes, that will still work

File: engine/lib/cache.php (hunk 6 above)
Comment (ewinslow) [inline engine/lib/cache.php]: done

**Comment by ewinslow** [comment]
@mrclay Thank you much for the thorough review

File: views/default/css/admin.php [hunk 8]
```diff
@@ -816,16 +816,16 @@
 	z-index:1;
 }
 .friends-picker-navigation-l {
-	background: url("<?php echo elgg_get_site_url(); ?>_graphics/friendspicker.png") no-repeat left top;
+	background: url("<?= elgg_get_simplecache_url("friendspicker.png"); ?>") no-repeat left top;
```
Comment (mrclay) [inline views/default/css/admin.php]: We don't allow `<?=` currently. I think we should, as it's always enabled now, and useful.

File: engine/classes/Elgg/Application.php (hunk 3 above)
Comment (mrclay) [inline engine/classes/Elgg/Application.php]: This needs to go now.

**Comment by ewinslow** [comment]
Retroactively fixed: https://github.com/Elgg/Elgg/pull/8473

File: engine/classes/Elgg/Application/CacheHandler.php [hunk 9]
```diff
@@ -199,22 +203,70 @@ protected function sendCacheHeaders($etag) {
...
 	 */
-	protected function sendContentType($view) {
-		$segments = explode('/', $view, 2);
-		switch ($segments[0]) {
-			case 'css':
-				header("Content-Type: text/css", true);
-				break;
+	protected function getContentType($view) {
+		
+		$fileType = $this->getViewFileType($view);
```
Comment (mrclay) [inline engine/classes/Elgg/Application/CacheHandler.php]: You want to set $extension here instead.

Comment (ewinslow) [inline engine/classes/Elgg/Application/CacheHandler.php]: Arg.... Thank you. This needs tests apparently :)

**Comment by ewinslow** [comment]
Fixed the `$fileType` vs `$extension` mismatch. Also added some test stubs.

**Comment by ewinslow** [comment]
SimpleCache is not in a place where it can be tested nicely yet. Still relying on too much global state.

**Comment by mrclay** [comment]
This feels pretty solid to me, though it certainly won't be as fast as direct file serving. Hopefully the lack of stuff like range requests won't be a problem.

**Comment by ewinslow** [comment]
I'm sure we can further optimize this to be faster/better. But for now yea,
this is definitely about dev usability over raw speed...

**Comment by ewinslow** [comment]
Did you want to see if we can get one more person to chime in?

**Comment by mrclay** [comment]
I'd certainly like to give @Elgg/owners a few days to chime in.

**Comment by ewinslow** [comment]
Ok, will wait until Tuesday. Monday seems to be our most active day, so
I'll assume no one has comments if I don't hear anything by then.

**Comment by hypeJunction** [comment]
This is really cool. I haven't read the entire code, but the approach is great. One wish though is to be able to reference static views from CSS without having to use PHP to get the simplecache URL.

**Comment by ewinslow** [comment]
@hypeJunction Wow, very glad to hear you're a fan of this approach. Reported your idea and my thoughts on it to https://github.com/Elgg/Elgg/issues/8486

**Comment by juho-jaakkola** [comment]
LGTM

File: engine/classes/Elgg/Application/CacheHandler.php [hunk 10]
```diff
@@ -200,22 +204,70 @@ protected function sendCacheHeaders($etag) {
...
+	private function getViewFileType($view) {
+		$extension = (new \SplFileInfo($view))->getExtension();
+		if ($extension) {
+			return $extension;
+		}
+		
+		if (preg_match('~(?:^|/)(css|js)(?:$|/)~', $view, $m)) {
+			return $m[1];
+		} else {
+			return 'unknown';
```
Comment (mrclay) [inline engine/classes/Elgg/Application/CacheHandler.php]: Wouldn't it be better to return empty string?

File: engine/lib/views.php [hunk 11]
```diff
@@ -1602,12 +1602,12 @@ function elgg_views_boot() {
...
 	elgg_register_simplecache_view('js/text.js');
 
-	elgg_register_js('elgg.require_config', elgg_get_simplecache_url('js', 'elgg/require_config'), 'head');
-	elgg_register_js('require', elgg_get_simplecache_url('js', 'require.js'), 'head');
-	elgg_register_js('jquery', elgg_get_simplecache_url('js', 'jquery.js'), 'head');
-	elgg_register_js('jquery-migrate', elgg_get_simplecache_url('js', 'jquery-migrate.js'), 'head');
-	elgg_register_js('jquery-ui', elgg_get_simplecache_url('js', 'jquery.ui.js'), 'head');
-	elgg_register_js('elgg', elgg_get_simplecache_url('js', 'elgg'), 'head');
+	elgg_register_js('elgg.require_config', elgg_get_simplecache_url('js/elgg/require_config'), 'head');
+	elgg_register_js('require', elgg_get_simplecache_url('js/require.js'), 'head');
```
Comment (mrclay) [inline engine/lib/views.php]: For each of these `elgg_get_simplecache_url` we can drop the associated `elgg_register_simplecache_view` call above.

File: engine/lib/views.php [hunk 12]
```diff
@@ -1616,12 +1616,12 @@ function elgg_views_boot() {
 	elgg_load_js('jquery-ui');
 	elgg_load_js('elgg');
 
-	$lightbox_js_url = elgg_get_simplecache_url('js', 'lightbox');
+	$lightbox_js_url = elgg_get_simplecache_url('js/lightbox');
 	elgg_register_js('lightbox', $lightbox_js_url);
 
-	elgg_register_css('lightbox', elgg_get_simplecache_url('js', 'lightbox.css'));
+	elgg_register_css('lightbox', elgg_get_simplecache_url('js/lightbox.css'));
```
Comment (mrclay) [inline engine/lib/views.php]: This is probably the conflict line. Sorry.

File: engine/classes/Elgg/Application/CacheHandler.php (hunk 10 above)
Comment (ewinslow) [inline engine/classes/Elgg/Application/CacheHandler.php]: I left it unchanged from where I moved it.

File: engine/lib/views.php (hunk 12 above)
Comment (ewinslow) [inline engine/lib/views.php]: Don't fret! not a big deal at all.

File: engine/lib/views.php (hunk 11 above)
Comment (ewinslow) [inline engine/lib/views.php]: Delete 4 superfluous calls.

**Comment by ewinslow** [comment]
I count 1 LGTM, 1 "This is really cool", and 1 "This feels pretty solid".

It's Tuesday. Pulling this in!

**Comment by iionly** [comment]
The move of the image files from _graphics to views/default is quite a BC breaking change. But this issue has no BC breaking label and there's also no mentioning about the consequences in the docs. Any plugin using $image = elgg_get_site_url() . '_graphics/<image_name> will fail to work on Elgg 2.0 if it isn't updated. Are there any plans to catch this problem to make the transition of plugins to 2.0 at least easier. If not, what is the suggested way to update the code of plugins?

**Comment by mrclay** [comment]
This is a really good point. As of last night we can put these images back in place and just point to them in `engine/views.php`.

**Comment by iionly** [comment]
I noticed that you mentioned the move of the images in https://github.com/Elgg/Elgg/pull/8540. Adding the images back in the old location (or linking them) would ease the transition. Though it might still be better to explain the recommended way of Elgg 2.0 for using the images, too.

Maybe "recommended plugin code updates" in addtion to "required plugin code updates" in docs/guides/upgrading.rst would help. The images issue could be one such recommendation. Another one would be the move to resources views. Most likely there are other "best practice" recommendations and it would help to know about them if you need to update your plugin code anyway.

**Comment by ewinslow** [comment]
> Any plugin using $image = elgg_get_site_url() . '_graphics/ will fail to
work on Elgg 2.0 if it isn't updated.

I added a redirect for this in .htaccess so that plugins referencing them
should still work.

However, if there are any direct filesystem references, those will fail.
That doesn't seem like as much of a concern though.

**Pull request merged**